src/alignment.jl

```diff
@@ -172,65 +209,68 @@ function check_alignment_anchors(anchors)
         if isdeleteop(op)
-            if anchors[i].seqpos != anchors[i-1].seqpos
-                error("Invalid anchor positions for reference deletion.")
+            if acur.seqpos != aprev.seqpos
+                error("Invalid anchor sequence positions for reference deletion.")
+            end
+            if acur.agnpos - aprev.agnpos != acur.refpos - aprev.refpos
+                error("Invalid anchor reference positions for reference deletion.")
             end
         # reference insertion operations
         elseif isinsertop(op)
-            if anchors[i].refpos != anchors[i-1].refpos
-                error("Invalid anchor positions for reference insertion.")
+            if acur.refpos != aprev.refpos
+                error("Invalid anchor reference positions for reference insertion.")
+            end
+            if acur.agnpos - aprev.agnpos != acur.seqpos - aprev.seqpos
+                error("Invalid anchor sequence positions for reference deletion.")
             end
         # match operations
         elseif ismatchop(op)
-            if anchors[i].refpos - anchors[i-1].refpos !=
-               anchors[i].seqpos - anchors[i-1].seqpos
-                error("Invalid anchor positions for match operation.")
+            if (acur.refpos - aprev.refpos != acur.seqpos - aprev.seqpos) ||
+               (acur.agnpos - aprev.agnpos != acur.seqpos - aprev.seqpos)
+               error("Invalid anchor positions for match operation.")
             end
         end
     end
 end
 
-# find the index of the first anchor that satisfies `i ≤ pos`
-@generated function findanchor(aln::Alignment, i::Integer, ::Type{Val{isseq}}) where isseq
-    pos = isseq ? :seqpos : :refpos
-    quote
-        anchors = aln.anchors
-        lo = 1
-        hi = lastindex(anchors)
-        if !(anchors[lo].$pos < i ≤ anchors[hi].$pos)
-            return 0
-        end
-        # binary search
-        @inbounds while hi - lo > 2
-            m = (lo + hi) >> 1
-            if anchors[m].$pos < i
-                lo = m
-            else  # i ≤ anchors[m].$pos
-                hi = m
-            end
-            # invariant (activate this for debugging)
-            #@assert anchors[lo].$pos < i ≤ anchors[hi].$pos
+# find the index of the first anchor that satisfies `i ≤ pos(anchor)`
+function findanchor(aln::Alignment, i::Integer, pos::Function)
+    anchors = aln.anchors
+    lo = 1
+    hi = lastindex(anchors)
+    @inbounds if !(pos(anchors[lo]) < i ≤ pos(anchors[hi]))
+        return 0
+    end
+    # binary search
+    @inbounds while hi - lo > 2
+        m = (lo + hi) >> 1
+        if pos(anchors[m]) < i
+            lo = m
+        else  # i ≤ anchors[m].$pos
+            hi = m
         end
-        # linear search
-        @inbounds for j in lo+1:hi
-            if i ≤ aln.anchors[j].$pos
-                return j
-            end
+        # invariant (activate this for debugging)
+        #@assert anchors[lo].$pos < i ≤ anchors[hi].$pos
```

Comment:
I think it should be still relevant, because the function logic was not touched, I've just added support for aligned position + made it a normal function, not `@generated`.
I'd rather keep it here as a comment: it nicely documents the constraint with the algorithm, but the context might be performance-critical to enable it.
I'll update it to use the current way of getting the position (`pos(anchors...))`) and check internally (without pushing here) whether it doesn't throw in tests.

Reply:
So I checked and tests do not throw if the line is enabled.